python/lsst/daf/butler/registries/sqlPreFlight.py

```diff
@@ -22,14 +22,83 @@
 __all__ = ("SqlPreFlight")
 
 import itertools
-import copy
-
 from sqlalchemy.sql import select, and_, text
 
-import lsst.log as lsstLog
+import lsst.log
+from lsst.sphgeom import Region
+from lsst.sphgeom.relationship import DISJOINT
+
+
+_LOG = lsst.log.Log.getLogger(__name__)
+
+
+def _scanDataUnits(dataUnits):
+    """Recursively scan units and their optional dependencies, return their names"""
+    for dataUnit in dataUnits:
+        yield dataUnit.name
+        yield from _scanDataUnits(dataUnit.optionalDependencies)
 
 
-_LOG = lsstLog.Log.getLogger(__name__)
+def _filterSummarizes(dataUnitJoins):
+    """Filter out DataUnitJoins that summarize other DataUnitJoins.
+
+    Parameters
+    ----------
+    dataUnitJoins : iterable of `DataUnitJoin`
+
+    Returns
+    -------
+    Iterator for DataUnitJoin which do not summarize any of the DataUnitJoins
+    in the input set.
+    """
+    dataUnitJoins = list(dataUnitJoins)
+    dataUnitJoinNames = set(join.name for join in dataUnitJoins)
+    for dataUnitJoin in dataUnitJoins:
+        summarizes = set(dataUnitJoin.summarizes or [])
+        # If it summarizes some other joins and all those joins are in the
+        # set of joins then we do not need it.
+        if summarizes and summarizes.issubset(dataUnitJoinNames):
+            continue
+        yield dataUnitJoin
+
+
+def _filterRegions(rowIter, firstRegionIndex):
+    """Fiters result rows that have non-overlapping regions.
+
+    Result set generated by query in selectDataUnits() method can include
+    set of regions in each row (encoded as bytes). Due to pixel-based
+    matching some regions may not overlap, this generator method filters
+    rows that have disjoint regions. If result row contains more than two
+    regions (this should not happen with our current schema) then row is
+    filtered if any of two regions are disjoint.
```

Comment:
Right, because coarser join tables summarize finer join tables you effectively end up with only one join table used in in SQL expression. True, this only applies for the case of single SkyMap/Camera, but this is all that pre-flight supports currently. I think that schema should more or less reflect expression system assumptions (with some extra things like `summarizes` defined in `schema.yaml`) so to me they are almost one and the same thing.